### PR TITLE
Better impl for char

### PR DESCRIPTION
**Pull request by Manishearth**
Fixes https://github.com/rust-fuzz/rust_arbitrary/issues/8

The String impl is still not good enough, since it blindly calls `char::from_arbitrary`, which doesn't work as well. We should do something slightly more UTF8 aware.

r? @fitzgen

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -188,20 +188,17 @@ impl_arbitrary_for_floats! {
...
-            if let Some(c) = ::std::char::from_u32(c) {
-                return Ok(c);
-            } else {
-                c -= 1;
-            }
+        use ::std::char;
+        const CHAR_END: u32 = 0x0011_000;
+        // The size of the surrogate blocks
+        const SURROGATES_START: u32 = 0xD800;
+        let mut c = <u32 as Arbitrary>::arbitrary(u)? % CHAR_END;
```
Comment (fitzgen) [inline src/lib.rs]: I think this should do the same trick that the derive does do reduce bias in enum discriminants, no?

Comment (Manishearth) [reply src/lib.rs]: I don't want to shift it, because that changes valid inputs.

i want to keep valid inputs unchanged

**Comment by Manishearth** [comment]
Hmm, another approach would be to parse as utf 8 instead, which might produce more success. That's trickier, though.

That would fix https://github.com/rust-fuzz/rust_arbitrary/issues/17 for free (aside from the container size)

**Comment by Manishearth** [comment]
Landing this anyway, this approach is strictly better.

**Pull request merged**